### PR TITLE
refactor(cli): replace ethers with viem

### PR DESCRIPTION
**Pull request by johngrantuk**
closes https://github.com/latticexyz/mud/issues/1006

### CODE REVIEW

File: packages/cli/src/utils/deploy.ts [hunk 1]
```diff
@@ -17,12 +16,27 @@ import { getRegisterTableCallData } from "./tables/getRegisterTableCallData";
 import { getTableIds } from "./tables/getTableIds";
 import { confirmNonce } from "./utils/confirmNonce";
 import { deployContract } from "./utils/deployContract";
-import { fastTxExecute } from "./utils/fastTxExecute";
+import { fastTx } from "./utils/fastTxExecute";
```
Comment (holic) [inline packages/cli/src/utils/deploy.ts:19]: Might see if we can use our internal createContract helper, since it has its own nonce tracking

Comment (johngrantuk) [reply packages/cli/src/utils/deploy.ts:19]: I did originally try that but came across issues when it was sending a bunch of txs. I could have been doing something wrong though so would definitely be interested to see how to do it correctly.

Comment (holic) [reply packages/cli/src/utils/deploy.ts:19]: oh interesting! thanks for trying, happy to go with this and revisit in the future

File: packages/cli/package.json [hunk 2]
```diff
@@ -57,6 +57,7 @@
     "path": "^0.12.7",
     "throttle-debounce": "^5.0.0",
     "typescript": "5.1.6",
+    "viem": "1.12.2",
```
Comment (holic) [inline packages/cli/package.json:60]: can we align this with the version we use everywhere else? `1.6.0`

File: packages/cli/src/utils/systems/utils.ts [hunk 3]
```diff
@@ -1,14 +1,26 @@
...
+type Param = AbiParameter & { components: any[] };
 
 export function loadFunctionSignatures(contractName: string, forgeOutDirectory: string): string[] {
   const { abi } = getContractData(contractName, forgeOutDirectory);
 
   return abi
     .filter((item) => ["fallback", "function"].includes(item.type))
     .map((item) => {
-      return `${item.name}${parseComponents(item.inputs)}`;
+      let name: string;
```
Comment (johngrantuk) [inline packages/cli/src/utils/systems/utils.ts:13]: I'm not sure if this is the best approach?

**Review by holic — DISMISSED** [review]

**Comment by johngrantuk** [comment]
> I just spent a bit trying to remove the `chain` requirement, because folks will often deploy with just an RPC URL, and they'll have no way to edit the `supportedChains` of the built CLI package.
> 
> While messing around with that, I turned on the `--debug` flag and discovered that we're getting _tons_ of nonce errors + retries with the approach in this PR. It does eventually seem to deploy properly (see passing tests), but I feel like we should handle nonces better, especially for non-local chains.
> 
> I think I'm gonna move on to working on #1611 and overhaul our deployment process. So I'll leave it to you to decide if you want to figure out nonce handling here or if we wanna just wait for handling this better in the declarative deployment approach.

No worries. TBH I'm not sure how its possible to use the manual nonce setting with Viem without this approach as I couldn't find a way to avoid the nonce errors (as far as I could see createContract was essentially doing the same but was just resubmitting with latest nonce instead of the original, which I think ends up causing issues when lots of transactions are being handled). I also tried with a que type setup where it would submit one after the other but this was way slower and AFAIK speed is one of the main priorities here.

**Pull request closed without merging**